packages/ra-core/src/auth/useRequiredAuth.ts

```diff
@@ -0,0 +1,155 @@
+ * @param {Object} options.params Any params you want to pass to the authProvider
+ * @param {Boolean} options.logoutOnFailure: Optional. Whether the user should be logged out if the authProvider fails to authenticate them. True by default.
+ * @param {Boolean} options.queryOptions: Optional. Options for react-query's useQuery hook.
+ *
+ * @returns The current auth check state. Destructure as { isPending }.
+ *
+ * @example
+ * import { useRequiredAuth, Loading } from 'react-admin';
+ *
+ * const MyPage = () => {
+ *     const { isPending } = useRequiredAuth();
+ *     if (isPending) {
+ *         return <Loading />;
+ *     }
+ *     // If not pending, the user is authenticated as they would be redirected otherwise
+ *     return <AuthenticatedContent />;
+ * };
+ */
+export const useRequiredAuth = <ErrorType = Error>({
+    params = emptyParams,
+    logoutOnFailure = true,
+    queryOptions = emptyParams,
+}: UseRequiredAuthOptions<ErrorType>): UseRequiredAuthResult<ErrorType> => {
+    const authProvider = useAuthProvider();
+    const logout = useLogout();
+    const basename = useBasename();
+    const navigate = useNavigate();
+
+    const queryResult = useQuery<boolean, any>({
+        queryKey: ['auth', 'checkAuth', params],
+        queryFn: ({ signal }) => {
+            // The authProvider is optional in react-admin
+            if (!authProvider) {
+                return true;
+            }
+            return authProvider
+                .checkAuth({ ...params, signal })
+                .then(() => true)
+                .catch(error => {
+                    // This is necessary because react-query requires the error to be defined
+                    if (error != null) {
+                        throw error;
+                    }
+
+                    throw new Error();
+                });
+        },
+        retry: false,
+        ...queryOptions,
+    });
+
+    useEffect(() => {
+        if (queryResult.error == null || queryResult.isFetching) return;
+        const loginUrl = removeDoubleSlashes(
+            `${basename}/${defaultAuthParams.loginUrl}`
+        );
+        if (logoutOnFailure) {
+            logout(
+                {},
+                queryResult.error && queryResult.error.redirectTo != null
+                    ? queryResult.error.redirectTo
+                    : loginUrl
+            );
+        } else {
+            navigate('/authentication-error');
+        }
+    }, [
+        basename,
+        logout,
+        logoutOnFailure,
+        navigate,
+        queryResult.error,
+        queryResult.isFetching,
+    ]);
+
+    const result = useMemo<UseRequiredAuthResult<ErrorType>>(() => {
+        return {
+            ...queryResult,
+            // If the data is undefined and the query isn't loading anymore, it means the query failed.
+            // In that case, we set authenticated to false unless there's no authProvider.
+            authenticated: queryResult.data,
```

Comment:
I think we shouldn't return anything except the default query result. Indeed, users will be redirected to either the login or the error page. @fzaninotto ?